A64-lrs-custom-metrics.md

```diff
@@ -0,0 +1,94 @@
+A64: xDS LRS Custom Metrics Support
+----
+* Author: yousukseung
+* Approver(s): markdroth
+* Status: Ready for Implementation
+* Implemented in: C-core
+* Last updated: 2023-05-10
+* Discussion at: TBA
+
+## Abstract
+
+This proposal describes how the gRPC xDS client will support custom metrics from the backends in load reports to the LRS server at the locality level.
+
+## Background
+
+The initial xDS functionality in the gRPC client described in [gRFC A27][A27] includes the ability to report load to the control plane via LRS. However, those load reports currently contain only basic request counts tracked by the client. [gRFC A51][A51] added support for backends to report custom backend metrics to the client via the [ORCA protocol][ORCA]. This proposal adds the ability for the client to propagate those backend metrics to the control plane in the LRS load reports.
+
+### Related Proposals:
+* [A27: xDS-Based Global Load Balancing][A27]
+* [A51: Custom Backend Metrics Support][A51]
+
+## Proposal
+
+### ORCA Integration
+
+The gRPC xDS client will include the entire `named_metrics` field in ORCA load reports from the backend. They will be considered application specific opaque values with no validation.
+
+### Per-Request Support Only
+
+Custom metrics in LRS load reports will only include `named_metrics` from per-request ORCA load reports and not OOB load reports. See [gRFC A51][A51] for more on per-request and OOB load reports.
+
+### Aggregation
+
+Each per-request ORCA load report will be associated with one request for the aggregation purpose. The value and count of each entry in `named_metrics` will be aggregated separately. All values will be considered cumulative and will be aggregated using addition. They will be tracked at the locality level.
+
+For example, following ORCA load reports
+```textproto
+// report 1/3
+rps_fractional: 10.0
+named_metrics { key: "key1" value: 1.0 }
+named_metrics { key: "key2" value: 2.0 }
+// report 2/3
+rps_fractional: 20.0
+named_metrics { key: "key2" value: 3.0 }
+named_metrics { key: "key3" value: 4.0 }
+// report 3/3
+rps_fractional: 30.0
+```
+will be aggreated to as follows.
+|field|value|number of requests|
+|------|---|---|
+|`key1`|1.0|1|
+|`key2`|5.0|2|
+|`key3`|4.0|1|
+
+### LRS Load Report
+
+The gRPC xDS client will include custom metrics in the `load_metric_stats` field in the locality stats. Each LRS report will include aggregated custom metrics with keys reported since the last report. Locally aggregated stats will be cleared and the associated total request counts will be reset to zero after each LRS load report is generated.
+
+An example LRS load report with custom metrics is as follows.
```

Comment:
Suggest making this example match up with the example in the previous section, so that it's clear how this maps to the aggregation.

Reply:
Done.